src/command.rs

```diff
@@ -1,36 +1,327 @@
-use clap::{AppSettings, Args, Parser, ValueEnum};
+use clap::{Args, Parser, Subcommand};
 
-#[derive(Parser)]
-#[clap(name = "cargo")]
-#[clap(bin_name = "cargo")]
+#[derive(Parser, Debug)]
+#[command(name = "cargo", bin_name = "cargo")]
 pub enum Cargo {
-    #[clap(name = "3ds")]
+    #[command(name = "3ds")]
     Input(Input),
 }
 
-#[derive(Args)]
-#[clap(about)]
-#[clap(global_setting(AppSettings::AllowLeadingHyphen))]
+#[derive(Args, Debug)]
+#[command(version, about)]
 pub struct Input {
-    #[clap(value_enum)]
-    pub cmd: CargoCommand,
-    pub cargo_opts: Vec<String>,
+    #[command(subcommand)]
+    pub cmd: CargoCmd,
 }
 
-#[derive(ValueEnum, Copy, Clone, Ord, PartialOrd, Eq, PartialEq)]
-pub enum CargoCommand {
-    Build,
-    Run,
-    Test,
-    Check,
-    Clippy,
+/// The cargo command to run. This command will be forwarded to the real
+/// `cargo` with the appropriate arguments for a 3DS executable.
+///
+/// If another command is passed which is not recognized, it will be passed
+/// through unmodified to `cargo` with RUSTFLAGS set for the 3DS.
+#[derive(Subcommand, Debug)]
+#[command(allow_external_subcommands = true)]
+pub enum CargoCmd {
+    /// Builds an executable suitable to run on a 3DS (3dsx).
+    Build(RemainingArgs),
+
+    /// Builds an executable and sends it to a device with `3dslink`.
+    Run(Run),
+
+    /// Builds a test executable and sends it to a device with `3dslink`.
+    ///
+    /// This can be used with `--test` for integration tests, or `--lib` for
+    /// unit tests (which require a custom test runner).
+    Test(Test),
+
+    // NOTE: it seems docstring + name for external subcommands are not rendered
+    // in help, but we might as well set them here in case a future version of clap
+    // does include them in help text.
+    /// Run any other `cargo` command with RUSTFLAGS set for the 3DS.
+    #[command(external_subcommand, name = "COMMAND")]
+    Passthrough(Vec<String>),
+}
+
+#[derive(Args, Debug)]
+pub struct RemainingArgs {
+    /// Pass additional options through to the `cargo` command.
+    ///
+    /// To pass flags that start with `-`, you must use `--` to separate `cargo 3ds`
+    /// options from cargo options. Any argument after `--` will be passed through
+    /// to cargo unmodified.
```

Comment:
Is this still the case?

Reply:
Unfortunately yes. I might try to revisit the wording of this or try adding some additional examples, perhaps, because it's not super intuititive. I also don't want to make the help output too verbose, though, so maybe some examples in the README would be better than dumping it all in the CLI help?

Reply:
Pushed some README changes plus reworded this help text, let me know what you think

Reply:
LGTM